EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -545,3 +289,308 @@ void disconnectContactors(uint8_t isContactorError)
+            break;
+
+        default: // Something went really wrong in the logic, stop everything
+            disconnectContactors(0);
+            *ptrState = SHUTDOWN;
+            return;
+    }
+
+    if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
+    {
+        return;
+    }
+
+    auxStatus.contactorError = contactorError;
+
+    if (*ptrHasDischargeBeenSet && *ptrHasChargeBeenSet)
+    {
+        auxStatus.startUpSequenceDone = 1;
+        *ptrState = DONE;
+    }
+
+    osMutexRelease(auxStatus.auxStatusMutex);
+
+    if (!*ptrHasChargeBeenSet)
+    {
+        *ptrPrevState = *ptrState;
+        *ptrState = CHARGE_ENABLE_CHECK;
+        HAL_GPIO_WritePin(CHARGE_ENABLE_GPIO_Port, CHARGE_ENABLE_Pin, GPIO_PIN_SET);
+    }
+    else if (!*ptrHasDischargeBeenSet)
+    {
+        *ptrPrevState = *ptrState;
+        *ptrState = CHARGE_ENABLE_CHECK;
+    }
+}
+
+void done()
+{
+    const uint8_t commonSense = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+    const uint8_t chargeSense = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
+    const uint8_t dischargeSense = !HAL_GPIO_ReadPin(DISCHARGE_SENSE_GPIO_Port, DISCHARGE_SENSE_Pin);
+
+    uint8_t isChargeTurningOn = 0;
+    uint8_t isDischargeTurningOn = 0;
+
+    if (orionStatus.shutOff)
+    {
+        *ptrState = SHUTDOWN;
+        return;
+    }
+    else if (!commonSense)
+    {
+        disconnectContactors(1);
+        *ptrState = CONTACTOR_DISCONNECTED;
+    }
+
+    if (orionStatus.allowCharge && orionStatus.chargeContactorOverriden && !chargeSense)
+    {
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.chargeContactorOverriden = 0;
+        osMutexRelease(orionStatus.orionStatusMutex);
+
+        isChargeTurningOn = 1;
+        *ptrHasChargeBeenSet = 0;
+        // Turn on charge Contactor
+        HAL_GPIO_WritePin(CHARGE_ENABLE_GPIO_Port, CHARGE_ENABLE_Pin, GPIO_PIN_SET);
+        *ptrPrevState = *ptrState;
+        *ptrState = CHARGE_ENABLE_CHECK;
+    }
+    else if (orionStatus.allowDischarge && orionStatus.dischargeContactorOverriden && !dischargeSense)
+    {
+        if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+        {
+            return;
+        }
+
+        orionStatus.dischargeContactorOverriden = 0;
+        osMutexRelease(orionStatus.orionStatusMutex);
+
+        isDischargeTurningOn = 1;
+        *ptrHasDischargeBeenSet = 0;
+        HAL_GPIO_WritePin(DISCHARGE_ENABLE_GPIO_Port, DISCHARGE_ENABLE_Pin, GPIO_PIN_SET);
+        *ptrPrevState = *ptrState;
+        *ptrState = DISCHARGE_ENABLE_CHECK;
+    }
+
+
+    // If charge contactor should be on, but isn't, disconnect
+    // or if discharge contactor should be on, but isn't, disconnect
+    if ((orionStatus.allowCharge && !chargeSense && !isChargeTurningOn && !orionStatus.chargeContactorOverriden) ||
```

Comment:
Same as above, but two variables. One for each condition on each side of the `||`